examples/DivisorCurve.jl

```diff
@@ -0,0 +1,327 @@
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.v)) || error("The point ", P, "is not on the curve.")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+#function Base.show(io::IO, D::ProjCurveDivisor)
+#    println(io, "On the curve ", D.C.eq, ", divisor defined by ", D.divisor)
+#end
+
+function ProjCurveDivisor(C::ProjPlaneCurve{S}, P::Oscar.Geometry.ProjSpcElem{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.v)) || error("The point is not on the curve.")
+    return ProjCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+# To get a nice output for the divisors.
+
+function AbstractAlgebra.needs_parentheses(A::Array{T, 1}) where T <: RingElem
+    return false
+end
+
+function AbstractAlgebra.expressify(@nospecialize(a::AffineCurveDivisor); context = nothing)
+   prod = Expr(:call, :+)
+   for (p, i) in a.divisor
+      ep = AbstractAlgebra.expressify(p.coord, context = context)
+      push!(prod.args, Expr(:call, :*, i, ep))
+   end
+   return prod
+end
+
+function AbstractAlgebra.needs_parentheses(A::Oscar.Geometry.ProjSpcElem{T}) where T <: RingElem
+    return false
+end
+
+function AbstractAlgebra.expressify(@nospecialize(a::ProjCurveDivisor); context = nothing)
+   prod = Expr(:call, :+)
+   for (p, i) in a.divisor
+      ep = AbstractAlgebra.expressify(p, context = context)
+      push!(prod.args, Expr(:call, :*, i, ep))
+   end
+   return prod
+end
+
+function Base.show(io::IO, ::MIME"text/plain", a::CurveDivisor)
+  print(io, AbstractAlgebra.obj_to_string(a, context = io))
+end
+
+function Base.show(io::IO, a::CurveDivisor)
+  print(io, AbstractAlgebra.obj_to_string(a, context = io))
+end
+
+################################################################################
+################################################################################
+
+function help_add(D::CurveDivisor, E::CurveDivisor)
+    D.C == E.C || error("The divisors are not on the same curve.")
+    F = copy(D.divisor)
+    G = copy(E.divisor)
+    ks = keys(F)
+    for P in keys(E.divisor)
+        if P in ks
+            a = F[P] + G[P]
+            F[P] = a
+            delete!(G, P)
+        end
+    end
+    for P in keys(G)
+        F[P] = G[P]
+    end
+    return F
+end
+
+################################################################################
+# Sum of two divisors
+
+function Base.:+(D::AffineCurveDivisor, E::AffineCurveDivisor)
+    F = help_add(D, E)
+    return AffineCurveDivisor(D.C, F)
+end
+
+function Base.:+(D::ProjCurveDivisor, E::ProjCurveDivisor)
+    F = help_add(D, E)
+    return ProjCurveDivisor(D.C, F)
+end
+
+################################################################################
+
+function Base.:-(D::CurveDivisor, E::CurveDivisor)
+    return D + -1*E
+end
```

Comment:
If one uses the above suggestion, then might as well replace this with the following slightly more efficient code:
```suggestion
function Base.:-(D::T, E::T) where {T <: CurveDivisor}
    _check_same_curve(D, E)
    return T(D.C, mergewith(-, D.divisor, E.divisor))
end
```

Reply:
I tried with mergewith(-, D.divisor, E.divisor), but if in E there are some points which are not in D, the corresponding coefficients do not have the right sign. I let it as it was before.

Reply:
Ah of course! My bad